File: Algo/Leetcode/139WordSplit.cpp
// dp. O(n^2) time, O(n) space
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        set<string> m;
        int n = s.length();
        for(auto w:wordDict) m.insert(w);
        vector<bool> dp(n+1, false);
        dp[0] = true;
        for(int i=1; i<=n; i++){
            for(int j=0; j<i; j++){
                if(dp[j] && m.find(s.substr(j, i-j))!=m.end()){
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[n];
    }
};

File: Algo/Leetcode/139WordSplit.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        wordSet = set(wordDict)
        n = len(s)
        dp = [False] * (n+1)
        dp[0] = True
        for i in range(1, n+1):
            for j in range(0, i):
                if s[j:i] in wordSet:
                    dp[i] = dp[i] or dp[j]  # either is fine

        return dp[-1]